pkg/liqo-controller-manager/namespace-controller/namespace_controller.go

```diff
@@ -1,147 +1,91 @@
+				PodOffloadingStrategy:    offv1alpha1.LocalAndRemotePodOffloadingStrategyType,
+				ClusterSelector:          corev1.NodeSelector{NodeSelectorTerms: []corev1.NodeSelectorTerm{}},
+			},
 		}
-		klog.Infof("Finalizer is correctly removed from namespace'%s'", namespace.GetName())
-
-		return ctrl.Result{}, nil
-	}
-
-	if !ctrlutils.ContainsFinalizer(namespace, namespaceControllerFinalizer) {
-		ctrlutils.AddFinalizer(namespace, namespaceControllerFinalizer)
-		if err := r.Patch(context.TODO(), namespace, client.Merge); err != nil {
-			klog.Errorf(" %s --> Unable to add finalizer on namespace '%s'", err, namespace.GetName())
+		if err := r.Create(ctx, namespaceOffloading); err != nil {
+			klog.Errorf("%s --> unable to Create NamespaceOffloading in namespace '%s'", err, namespace.Name)
 			return ctrl.Result{}, err
 		}
 	}
 
-	// 1. If mapping.liqo.io label is not present there are no remote namespaces associated
-	// with this namespace, removeMappings is full
-	if remoteNamespaceName, ok := namespace.GetLabels()[mappingLabel]; ok {
-		// 2.a If offloading.liqo.io is present there are remote namespaces on all virtual nodes
-		if _, ok = namespace.GetLabels()[offloadingLabel]; ok {
-			klog.Infof(" Offload namespace '%s' on all remote clusters", namespace.GetName())
-			if err := r.addDesiredMappings(namespace, remoteNamespaceName, removeMappings); err != nil {
-				return ctrl.Result{}, err
-			}
-
-			for k := range removeMappings {
-				delete(removeMappings, k)
-			}
-		} else {
-			// 2.b Iterate on all virtual nodes' labels, if the namespace has all the requested labels, is necessary to
-			// offload it onto remote cluster associated with the virtual node
-			nodes := &corev1.NodeList{}
-			if err := r.List(context.TODO(), nodes,
-				client.MatchingLabels{liqoconst.TypeLabel: liqoconst.TypeNode}); err != nil {
-				klog.Error(err, " --> Unable to List all virtual nodes")
-				return ctrl.Result{}, err
-			}
-
-			if len(nodes.Items) == 0 {
-				klog.Info(" No VirtualNode at the moment")
-				return ctrl.Result{}, nil
-			}
-
-			for i := range nodes.Items {
-				if checkOffloadingLabels(namespace, &nodes.Items[i]) {
-					if err := r.addDesiredMapping(namespace, remoteNamespaceName,
-						removeMappings[nodes.Items[i].Annotations[liqoconst.RemoteClusterID]]); err != nil {
-						return ctrl.Result{}, err
-					}
-					delete(removeMappings, nodes.Items[i].Annotations[liqoconst.RemoteClusterID])
-					klog.Infof(" Offload namespace '%s' on remote cluster: %s", namespace.GetName(),
-						nodes.Items[i].Annotations[liqoconst.RemoteClusterID])
-				}
-			}
-		}
-	}
-
-	if len(removeMappings) > 0 {
-		klog.Info(" Delete all unnecessary entries in NamespaceMaps")
-		if err := r.removeDesiredMappings(namespace.GetName(), removeMappings); err != nil {
+	// check if enabling Liqo label is removed
+	if !liqoLabelPresence(namespace.Labels) && namespaceOffloadingIsPresent {
+		if err := r.Delete(ctx, namespaceOffloading); err != nil {
```

Comment:
@giorio94 @Andreagit97 I was wondering: should we ensure that this controller has ownership of the resource before deleting it? In other words, should the controller not remove any resource it was not created by him? I personally would say yes, since the presence of the label is optional and the labeling when a NamespaceOffloading is already present won't have any effect.

Reply:
sure, I will do it, thank you for the advise.